prism/prism.c

```diff
@@ -16960,22 +16960,30 @@ parse_strings(pm_parser_t *parser, pm_node_t *current, bool accepts_label, uint1
 #define PM_PARSE_PATTERN_TOP 1
 #define PM_PARSE_PATTERN_MULTI 2
 
+typedef struct {
+    bool in_alternative_pattern;
+    bool capture_in_pattern;
+} pm_pattern_capturing_t;
+
 static pm_node_t *
-parse_pattern(pm_parser_t *parser, pm_constant_id_list_t *captures, uint8_t flags, pm_diagnostic_id_t diag_id, uint16_t depth);
+parse_pattern(pm_parser_t *parser, pm_constant_id_list_t *captures, uint8_t flags, pm_diagnostic_id_t diag_id, uint16_t depth, pm_pattern_capturing_t *capturing);
 
 /**
  * Add the newly created local to the list of captures for this pattern matching
  * expression. If it is duplicated from a previous local, then we'll need to add
  * an error to the parser.
  */
 static void
-parse_pattern_capture(pm_parser_t *parser, pm_constant_id_list_t *captures, pm_constant_id_t capture, const pm_location_t *location) {
+parse_pattern_capture(pm_parser_t *parser, pm_constant_id_list_t *captures, pm_constant_id_t capture, const pm_location_t *location, pm_pattern_capturing_t *capturing) {
     // Skip this capture if it starts with an underscore.
     if (*location->start == '_') return;
 
     if (pm_constant_id_list_includes(captures, capture)) {
         pm_parser_err(parser, location->start, location->end, PM_ERR_PATTERN_CAPTURE_DUPLICATE);
+    } else if (capturing->in_alternative_pattern) {
```

Comment:
prism can parse multiple ruby versions and the prism gem does get backported to stable branches.
Since this practically already is a syntax error I am unsure of how important that actually is, but I did want to mention it. Same below.

```suggestion
    } else if (capturing->in_alternative_pattern && parser->version >= PM_OPTIONS_VERSION_CRUBY_3_5) {
```